servers/src/main/java/tachyon/master/file/FileSystemMaster.java

```diff
@@ -104,6 +106,134 @@
+
+    public TTLBucket(long startTimeMs) {
+      mTTLIntervalStartTimeMs = startTimeMs;
+      mFiles = new LinkedList<InodeFile>();
+    }
+
+    public long getTTLIntervalStartTimeMs() {
+      return mTTLIntervalStartTimeMs;
+    }
+
+    public void setTTLIntervalStartMs(long startTimeMs) {
+      mTTLIntervalStartTimeMs = startTimeMs;
+    }
+
+    public long getTTLIntervalEndTimeMs() {
+      return mTTLIntervalStartTimeMs + mTTLCheckerIntervalMs;
+    }
+
+    public List<InodeFile> getFiles() {
+      return mFiles;
+    }
+
+    public void addFile(InodeFile file) {
+      mFiles.add(file);
+    }
+  }
+
+  /**
+   * A list of non-empty {@link TTLBucket}s sorted by ttl interval start time of each bucket.
+   *
+   * Two adjacent buckets may not have adjacent intervals since there may be no files with ttl value
+   * in the skipped intervals.
+   *
+   * Thread-safety is guaranteed by {@link ConcurrentSkipListSet}.
+   */
+  private class TTLBucketList {
+    /**
+     * List of buckets sorted by interval start time.
+     * SkipList is used for O(logn) insertion and retrieval, see {@link ConcurrentSkipListSet}.
+     */
+    private ConcurrentSkipListSet<TTLBucket> mBucketList;
+
+    public TTLBucketList() {
+      mBucketList = new ConcurrentSkipListSet<TTLBucket>(new Comparator<TTLBucket>() {
+        @Override
+        public int compare(TTLBucket ttlBucket, TTLBucket ttlBucket2) {
+          long startTime1 = ttlBucket.getTTLIntervalStartTimeMs();
+          long startTime2 = ttlBucket2.getTTLIntervalStartTimeMs();
+          if (startTime1 < startTime2) {
+            return -1;
+          }
+          if (startTime1 == startTime2) {
+            return 0;
+          }
+          return 1;
+        }
+      });
+    }
+
+    /**
+     * Inserts an {@link InodeFile} to the appropriate bucket where its ttl end time lies in the
+     * bucket's interval, if no appropriate bucket exists, a new bucket will be created to contain
+     * this file, if ttl value is {@link Constants#NO_TTL}, the file won't be inserted to any
+     * buckets and nothing will happen.
+     *
+     * @param file the file to be inserted
+     */
+    public void insert(InodeFile file) {
+      if (file.getTTL() == Constants.NO_TTL) {
+        return;
+      }
+
+      long ttlEndTimeMs = file.getCreationTimeMs() + file.getTTL();
+      // Gets the last bucket with interval start time less than or equal to the file's life end
+      // time.
+      TTLBucket bucket = mBucketList.floor(new TTLBucket(ttlEndTimeMs));
+      if (bucket == null || bucket.getTTLIntervalEndTimeMs() <= ttlEndTimeMs) {
+        // 1. There is no bucket in the list.
+        // 2. All buckets' interval start time is larger than the file's life end time.
+        // 3. No bucket actually contains ttlEndTimeMs in its interval.
+        // So a new bucket should should be added with ttlEndTimeMs as its interval start.
+        bucket = new TTLBucket(ttlEndTimeMs);
```

Comment:
I should set a better interval start other than `ttlEndTimeMs`, thanks for hint.